File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        lpar = ('(', '{', '[')
        stack = [-99]
        
        for c in s:
            if c in lpar:
                stack.append(c)
            else:
                if c == ')' and stack[-1] == '(':
                    stack.pop()
                elif c == '}' and stack[-1] == '{':
                    stack.pop()                
                elif c == ']' and stack[-1] == '[':
                    stack.pop()
                else:
                    return False
        return len(stack) == 1
                    
                    
        

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {

    public boolean isValid(String s) {
        Stack<Character> st = new Stack<Character>();
        st.push('$');
        char[] c = s.toCharArray();

        for (int i = 0; i < s.length(); i++) {
            if (c[i] == '(' || c[i] == '{' || c[i] == '[') {
                st.push(c[i]);
                continue;
            }

            if (c[i] == ')' && st.peek() == '(') {
                st.pop();
            } else if (c[i] == '}' && st.peek() == '{') {
                st.pop();
            } else if (c[i] == ']' && st.peek() == '[') {
                st.pop();
            } else {
                return false;
            }
        }

        if (st.peek() == '$') {
            return true;
        } else {
            return false;
        }
    }
}
